jaxrs/src/main/java/org/killbill/billing/jaxrs/resources/CatalogResource.java

```diff
@@ -118,9 +118,28 @@ public CatalogResource(final JaxrsUriBuilder uriBuilder,
     @Produces(APPLICATION_XML)
     @ApiOperation(value = "Retrieve the full catalog as XML", response = String.class, hidden = true)
     @ApiResponses(value = {})
-    public Response getCatalogXml(@javax.ws.rs.core.Context final HttpServletRequest request) throws Exception {
+    public Response getCatalogXml(@QueryParam(QUERY_REQUESTED_DT) final String requestedDate,
+                                  @javax.ws.rs.core.Context final HttpServletRequest request) throws Exception {
         final TenantContext tenantContext = context.createTenantContextNoAccountId(request);
-        return Response.status(Status.OK).entity(XMLWriter.writeXML((VersionedCatalog) catalogUserApi.getCatalog(catalogName, tenantContext), VersionedCatalog.class)).build();
+        final DateTime catalogDateVersion = requestedDate != null ?
+                                            DATE_TIME_FORMATTER.parseDateTime(requestedDate).toDateTime(DateTimeZone.UTC) :
+                                            null;
+
+        final VersionedCatalog catalog = (VersionedCatalog) catalogUserApi.getCatalog(catalogName, tenantContext);
+
+        String result = null;
+        if (catalogDateVersion != null) {
+            for (final StandaloneCatalog v : catalog.getVersions()) {
+                if (v.getEffectiveDate().compareTo(catalogDateVersion.toDate()) == 0) {
```

Comment:
I am wondering if we should not use `v.getEffectiveDate().compareTo(catalogDateVersion.toDate()) >= 0`, otherwise this forces the user to know exactly the timestamp of each catalog. Let 's assume two versions: `2018-01-01T00:00:00+00:00` and `2018-03-01T00:00:00+00:00`:

* Providing any date >=  `2018-01-01T00:00:00+00:00` and < `2018-03-01T00:00:00+00:00` would return first version
* Providing any date >= `2018-03-01T00:00:00+00:00` would return second version.

That way the parameter request date represents the date within the range of date for which this catalog version is effective.

@pierre Any thoughts?

Reply:
@sbrossie I had indeed thought about it. However, in this case, versions are used as a number or `id`: we are looking for a specific catalog version here (it's different than retrieving the catalog at a given point in time).

That being said, I'm happy either way - your call.

Reply:
Agreed that for the KAUI use cases, where exact versions -- as specified with their `effectiveDate` -- this works 'as is', but thinking of api, might be more powerful to all for any date within the range ?